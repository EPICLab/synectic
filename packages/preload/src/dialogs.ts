import type {SaveDialogOptions, OpenDialogOptions} from 'electron';
import {ipcRenderer} from 'electron';

export const fileOpen = async (
  options: OpenDialogOptions,
): Promise<Electron.OpenDialogReturnValue> => await ipcRenderer.invoke('fileOpenDialog', options);

export const fileSave = async (
  options: SaveDialogOptions,
): Promise<Electron.SaveDialogReturnValue> => await ipcRenderer.invoke('fileSaveDialog', options);
