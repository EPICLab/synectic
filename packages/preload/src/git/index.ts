export * from './git-add';
export * from './git-branch';
export * from './git-cat-file';
export * from './git-check-ref-format';
export * from './git-checkout';
export * from './git-clone';
export * from './git-commit';
export * from './git-config';
export * from './git-ignore';
export * from './git-log';
export * from './git-merge-base';
export * from './git-merge';
export * from './git-path';
export * from './git-remote';
export * from './git-reset';
export * from './git-restore';
export * from './git-rev-list';
export * from './git-rev-parse';
export * from './git-show-branch';
export * from './git-status';
export * from './git-utils';
export * from './git-version';
export * from './git-worktree';
