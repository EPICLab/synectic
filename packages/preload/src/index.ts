/**
 * @module preload
 */

export {fileOpen, fileSave} from './dialogs';
export {flattenArray, flattenObject} from './flatten';
export {
  add,
  catFile,
  checkGitVersion,
  checkRefFormat,
  checkUnmergedBranch,
  checkUnmergedPath,
  checkoutBranch,
  checkoutDetached,
  checkoutNewBranch,
  checkoutPathspec,
  cloneRepo,
  commit,
  copyBranch,
  createBranch,
  deleteBranch,
  errorLogger,
  extractFromURL,
  extractRepoName,
  fetchMergingBranches,
  fileStatus,
  filterIgnore,
  getBranchRoot,
  getConfig,
  getCredentials,
  getIgnore,
  getMatchUp,
  getRemote,
  getRemoteConfig,
  getRoot,
  getWorktreePaths,
  isValidRepository,
  isValidRepositoryURL,
  listBranch,
  log,
  mergeBase,
  mergeBranch,
  mergeInProgress,
  moveBranch,
  processBranchOutput,
  processCatFileOutput,
  processLogOutput,
  processMergeBaseOutput,
  processMergeOutput,
  processRemoteOutput,
  processShowBranchOutput,
  processStatusCode,
  processWorktreeOutput,
  reset,
  resolveURL,
  restore,
  revList,
  revParse,
  revParseArgs,
  setConfig,
  setUpstreamBranch,
  showBranch,
  unsetUpstreamBranch,
  worktreeAdd,
  worktreeList,
  worktreeLock,
  worktreeMove,
  worktreePrune,
  worktreeRemove,
  worktreeRepair,
  worktreeStatus,
  worktreeUnlock,
  type Conflict,
  type GitConfig,
  type MergeAction,
  type WorktreePaths,
} from './git';
export {
  decompressBinaryObject,
  execute,
  extractDirname,
  extractExtension,
  extractFilename,
  extractStats,
  filterReadArray,
  getDescendantPaths,
  isDescendant,
  isDirectory,
  isEqualPaths,
  readDirAsync,
  readDirAsyncDepth,
  readFileAsync,
  replaceExt,
  validateFileName,
  writeFileAsync,
} from './io';
export {
  clipboard,
  cwd,
  dirname,
  join,
  normalize,
  openExternal,
  platform,
  relative,
  sep,
  sha256sum,
  uuid,
} from './platform';
export {
  asyncFilter,
  delay,
  deserialize,
  equalArrayBuffers,
  equalArrays,
  equalMaps,
  filterObject,
  filteredArrayEquality,
  filteredObjectEquality,
  getConflictingChunks,
  getRandomInt,
  hasUpdates,
  isDefined,
  isFilled,
  isModified,
  isNumber,
  isPresent,
  isStaged,
  isUnmerged,
  isWrappedQuote,
  objectEntries,
  objectifyPath,
  partition,
  removeDuplicates,
  removeNullableProperties,
  removeObjectProperty,
  removeUndefined,
  symmetrical,
  toArrayBuffer,
} from './utils';
export {versions} from './versions';
