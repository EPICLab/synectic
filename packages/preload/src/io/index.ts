export * from './io-decompress';
export * from './io-exec';
export * from './io-extension';
export * from './io-extractName';
export * from './io-extractStats';
export * from './io-filterReadArray';
export * from './io-getDescendants';
export * from './io-isDescendant';
export * from './io-isDirectory';
export * from './io-isEqualPaths';
export * from './io-readDir';
export * from './io-readFile';
export * from './io-validateFileName';
export * from './io-writeFile';

/**
 * **WARNING** If a method is not included here, please see
 * [node-fs-extra](https://github.com/jprichardson/node-fs-extra) for possible utility functions
 * that improve upon the default filesystem functions in [Node.js FS
 * module](https://nodejs.org/api/fs.html). If there is no functions in `node-fs-extra` or `fs`,
 * then a new method can be added here.
 */
