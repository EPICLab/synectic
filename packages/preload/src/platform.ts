import {type BinaryLike, createHash, randomUUID} from 'node:crypto';
import {shell} from 'electron';

export {cwd, platform} from 'node:process';
export {dirname, join, relative, normalize, sep} from 'node:path';
export {clipboard} from 'electron';

export const sha256sum = (data: BinaryLike) => {
  return createHash('sha256').update(data).digest('hex');
};

export const uuid = () => randomUUID();

export const openExternal = (url: string) => shell.openExternal(url);
