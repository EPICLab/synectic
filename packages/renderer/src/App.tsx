/// <reference types="vite-plugin-svgr/client" />

import type {Filetype} from '@syn-types/filetype';
import {useEffect} from 'react';
import {Provider} from 'react-redux';
import {AppThemeProvider} from './components/AppTheme';
import Canvas from './components/Canvas';
import filetypesJson from './containers/filetypes.json';
import redux from './store/store';
import {importFiletypes} from './store/thunks/filetypes';

const App = () => {
  useEffect(() => {
    redux.store.dispatch(importFiletypes(filetypesJson as Omit<Filetype, 'id'>[])); // load all supported filetype handlers into Redux store
  }, []); // run the effect only once; after the first render

  return (
    <Provider store={redux.store}>
      <AppThemeProvider>
        <Canvas />
      </AppThemeProvider>
    </Provider>
  );
};
export default App;
