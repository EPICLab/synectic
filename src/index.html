<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>Synectic Development Environment</title>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>
