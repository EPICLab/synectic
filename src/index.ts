import { app, BrowserWindow, dialog, ipcMain } from 'electron';
import installExtension, { REDUX_DEVTOOLS, REACT_DEVELOPER_TOOLS } from 'electron-devtools-installer';
// This allows TypeScript to pick up the magic constant that's auto-generated by Forge's Webpack
// plugin that tells the Electron app where to look for the Webpack-bundled app code (depending on
// whether you're running in development or production).
declare const MAIN_WINDOW_WEBPACK_ENTRY: string;

// Handle creating/removing shortcuts on Windows when installing/uninstalling.
if (require('electron-squirrel-startup')) { // eslint-disable-line global-require
  app.quit();
}

const createWindow = (): void => {
  // Create the browser window.
  const mainWindow = new BrowserWindow({
    height: 1200,
    width: 1400,
    webPreferences: {
      contextIsolation: false,
      nodeIntegration: true,
      webviewTag: true
    }
  });

  // and load the index.html of the app.
  mainWindow.loadURL(MAIN_WINDOW_WEBPACK_ENTRY);

  // Open the DevTools.
  if (process.env.NODE_ENV !== 'production') {
    mainWindow.webContents.openDevTools();
  }
};

// This method will be called when Electron has finished
// initialization and is ready to create browser windows.
// Some APIs can only be used after this event occurs.
app.on('ready', createWindow);
app.whenReady().then(() => {
  installExtension([REDUX_DEVTOOLS, REACT_DEVELOPER_TOOLS])
    .then((name) => console.log(`Added extension: ${name}`))
    .catch((err) => console.log('An error occurred: ', err))
});

// Handle renderer requests for Electron `showOpenDialog` and returns filepath results.
ipcMain.handle('fileOpenDialog', async (_event, properties) => {
  return await dialog.showOpenDialog({ properties: [...properties, 'multiSelections'] });
});

// Handle renderer requests for Electron `showSaveDialog` and returns the results value.
ipcMain.handle('fileSaveDialog', async (_event, { defaultPath, buttonLabel, properties }) => {
  return await dialog.showSaveDialog({ defaultPath, buttonLabel, properties });
});

// Quit when all windows are closed, except on macOS. There, it's common
// for applications and their menu bar to stay active until the user quits
// explicitly with Cmd + Q.
app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

app.on('activate', () => {
  // On OS X it's common to re-create a window in the app when the
  // dock icon is clicked and there are no other windows open.
  if (BrowserWindow.getAllWindows().length === 0) {
    createWindow();
  }
});

// In this file you can include the rest of your app's specific main process
// code. You can also put them in separate files and import them here.
