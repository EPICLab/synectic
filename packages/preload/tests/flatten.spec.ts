import {describe, expect, test} from 'vitest';
import {flattenArray, flattenObject} from '../src';

describe('flattenArray', () => {
  test('flattenArray resolves simple arrays of depth 1', () => {
    expect(flattenArray([1])).toStrictEqual([1]);
  });

  test('flattenArray resolves arrays of equal max-depths', () => {
    expect(flattenArray([[1], [2]])).toStrictEqual([1, 2]);
  });

  test('flattenArray resolves arrays of non-equal max-depths', () => {
    expect(flattenArray([[1], 2])).toStrictEqual([1, 2]);
  });

  test('flattenArray resolves heterogenous arrays of primitive types', () => {
    expect(
      flattenArray([
        [1, 2],
        [['apple'], [true]],
      ]),
    ).toStrictEqual([1, 2, 'apple', true]);
  });

  test('flattenArray resolves heterogenous arrays of object types', () => {
    const flatArray = [{id: 1}, {2: 'pink'}, 'apple', {state: true}];
    expect(
      flattenArray([
        [{id: 1}, {2: 'pink'}],
        [['apple'], [{state: true}]],
      ]),
    ).toStrictEqual(flatArray);
  });
});

describe('flattenObject', () => {
  test('flattenObject resolves nested heterogenous JavaScript objects', () => {
    const origObj = {a: 'string', b: {c: 13}};
    expect(flattenObject(origObj)).toMatchSnapshot();
  });
});
