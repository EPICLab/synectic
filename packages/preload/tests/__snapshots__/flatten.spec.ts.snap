// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flattenObject > flattenObject resolves nested heterogenous JavaScript objects 1`] = `
{
  "a": "string",
  "c": 13,
}
`;
